File: app/src/main/java/ali/khaleghi/batman/service/repository/interceptors/OfflineInterceptor.java
package ali.khaleghi.batman.service.repository.interceptors;

import ali.khaleghi.batman.util.AppConfigs;
import ali.khaleghi.batman.util.Toaster;
import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.util.Log;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;

public class OfflineInterceptor implements Interceptor {

    private Context context;

    public OfflineInterceptor(Context context) {
        this.context = context;
    }

    @NotNull
    @Override
    public Response intercept(@NotNull Chain chain) throws IOException {
        Request request = chain.request();

        if (!isOnline()) {
            int maxStale = 60 * 60 * AppConfigs.CACHE_EXPIRE_TIME;
            request = request.newBuilder()
                    .header("Cache-Control", "public, only-if-cached, max-stale=" + maxStale)
                    .removeHeader("Pragma")
                    .build();
        }

        return chain.proceed(request);
    }

    private boolean isOnline() {
        ConnectivityManager cm = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo netInfo = cm.getActiveNetworkInfo();
        return netInfo != null && netInfo.isConnectedOrConnecting();
    }
}

File: app/src/main/java/ali/khaleghi/batman/service/repository/interceptors/OfflineInterceptor.kt
package ali.khaleghi.batman.service.repository.interceptors

import ali.khaleghi.batman.util.AppConfigs
import ali.khaleghi.batman.util.NetworkState
import ali.khaleghi.batman.util.Toaster
import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkInfo
import android.util.Log
import okhttp3.CacheControl
import okhttp3.Interceptor
import okhttp3.Request
import okhttp3.Response

import java.io.IOException

class OfflineInterceptor(private val context: Context) : Interceptor {

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        var request = chain.request()

        if (!NetworkState.isOnline(context)) {

            val maxStale = 60 * 60 * AppConfigs.CACHE_EXPIRE_TIME
            request = request.newBuilder()
                .header("Cache-Control", "public, only-if-cached, max-stale=$maxStale")
                .removeHeader("Pragma")
                .build()
        }

        return chain.proceed(request)
    }

}
